### ISSUE TITLE
JIT: Generating rep movsb for cpblk with non-constant arguments

### ISSUE THREAD

**Issue by jamesqo** [dotnet/coreclr#6661]
Separated out from the discussion in #6638

Currently the JIT makes a helper call to `JIT_MemCpy` for non-constant sizes. If it could generate `rep movsb` inline instead for amd64/x86, that would be much more efficient. Doing this would allow for a better implementation of `Buffer.MemoryCopy`, which right now has xmm as the best thing available.

cc @jkotas @nietras @mikedn @tannergooding

**Comment by benaadams** [comment]
Does it use this for x86 and x64? https://github.com/dotnet/coreclr/blob/master/src/vm/amd64/CrtHelpers.asm

**Comment by benaadams** [comment]
CPUID EAX=7, EBX=0 => Bit 9 Enhanced REP MOVSB/STOSB

**Comment by redknightlois** [comment]
For the fastest memcopy working on almost every any architecture, you should have to check for the cpuid and relink using an stub on load. Then use the fastest method and instruction set for the executing architecture. That is probably something a JIT helper could do.

**Comment by benaadams** [comment]
> For the fastest memcopy working on almost every any architecture, you should have to check for the cpuid and relink using an stub on load. Then use the fastest method and instruction set for the executing architecture.

Lots of variant code bases, might only want to target a few common scenarios. At least it seems to be standardising.

**Comment by jamesqo** [comment]
@benaadams It's [this part](https://github.com/dotnet/coreclr/blob/master/src/jit/codegenxarch.cpp#L2815-L2827) in the code generator and [here](https://github.com/dotnet/coreclr/blob/8e0cbbba5b41744873eb6263069497a5dbdfedd4/src/jit/lowerxarch.cpp#L1673-L1762) in the lowerer; currently, it looks as if for AMD64 it's generating a couple of `xmm` writes for constants divisible by 16 under 64, and otherwise generating a `call`.

In fact, right now I don't the JIT *ever* generates a `rep movsb` for x64, even for constant values; [this assert](https://github.com/dotnet/coreclr/blob/master/src/jit/codegenxarch.cpp#L3473) will clearly always fail, since if you do a GitHub search CPBLK_UNROLL_LIMIT is 64 and CPBLK_MOVS_LIMIT is 16. Perhaps there was a typo in [the latter's definition](https://github.com/dotnet/coreclr/blob/master/src/jit/target.h#L699-L704)?

cc @BruceForstall

**Comment by redknightlois** [comment]
@benaadams not really. Even though there are a few variants, even the standard prefetch/non-temporal version should be unbeatable against any potential memcopy written in managed code at any size.

**Comment by tannergooding** [comment]
@redknightlois, that may be true when looking at a purely unsafe `FCALL` to `memcpy`. However, when you have to do a `QCALL` and do argument validation and exception handling, the managed implementation beats out the unmanaged implementation for small buffers (`QCALL`s are expensive).

Additionally, looking at the `memcpy` implementation on Windows, the `prefetch`/`non-temporal` version is only used when the number of 128-byte blocks to copy is greater than `8239` (`1,054,592` bytes). Feel free to look at the disassembly yourself to confirm 😄 

The Intel Optimization Manual (http://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-optimization-manual.html) has some great documentation on the exact benefits and drawbacks of various implementations and even provides heuristics for how to implement `memcpy` using `xmm` and when to use prefetching and non-temporal stores.

**Comment by redknightlois** [comment]
@tannergooding I know, we have our own memory copy implementation since 1.5 years ago (that's why I a following this thread closely). Any clever improvement in ```memcpy``` (even if not "officially aproved" for inclusion in CoreCLR) would have a real payoff for us.  

And as you mention that is the point, we are still thinking in terms of ```QCALL``` instead of implementing it at the JIT/CoreCLR level as a ```FCALL```.  The potential improvement of doing that straight as a code-generation solution is too big to pass. Also handcoding your ```memcpy``` implementation to be "faster" based on what the JIT can recognize and optimize today is a very brittle solution for such a core function. We benchmark our ```memcpy```  algorithm every 6 months on multiple hardware and also JIT versions to be sure that we are not regressing when optimizations and/or changes to the code generation happen. We pay the cost of finding the 'sweet spot' as we cannot control the environment our code will run (something at jitting time we can know), but we would certainly would like to avoid it altogether.

**Comment by benaadams** [comment]
Although would the linux version be any better as that just calls the platform versions https://github.com/dotnet/coreclr/blob/master/src/vm/amd64/crthelpers.S#L16-L38

**Comment by redknightlois** [comment]
@benaadams it is a managed version. So yes, I believe doing all this stuff there with proper CPU dispatching should be the way to go. I have a few bigger fishes to fry on our side regarding IO, but if the time comes and we need to make ```memcpy``` faster there would probably be time to invest in some assembler work. Question that has been bugging me for a long time already is: when that one is selected when jitting?

**Comment by benaadams** [comment]
is cpblk so copying structs as arguments and for assignment I think? or [`Unsafe.CopyBlock`](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il#L122-L131)

**Comment by Anderman** [comment]
movsb has some overhead see [SO](http://stackoverflow.com/questions/33902068/what-setup-does-rep-do/33905887). 

I created a test with rep movsb. movsb was called by pinvoke. The movsb test is corrected by PInvoke overhead. Arraycopy also use movsb for size>128. I used vector copy

![image](https://cloud.githubusercontent.com/assets/1858745/17986224/fa5accee-6b19-11e6-8e3b-c6c042ee8f0f.png)

**Comment by tannergooding** [comment]
Possibly: https://github.com/dotnet/designs/issues/13

**Comment by tannergooding** [comment]
Based on https://github.com/dotnet/coreclr/issues/25505, there looks to be a lot more considerations needed for AMD machines that may make this optimization undesirable.
